File: web/mocks/handlers.ts
import { http, HttpResponse, delay } from "msw";
import {
  STATE,
  balanceFrame,
  fundingFrame,
  nowNs,
  nsStr,
  positionsFrame,
  pushFill,
  recentPublicTrades,
  refreshMarks,
  snapshotOf,
} from "./data/seed";
import type { Fill, MarketId, OpenOrder, OrderRequest } from "@/lib/types/contract";

const BASE = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8080";

function u(path: string) {
  return `${BASE}${path}`;
}

function known(marketId: string): marketId is MarketId {
  return marketId === "btc-usd" || marketId === "eth-usd" || marketId === "sol-usd";
}

export const handlers = [
  http.get(u("/v1/markets"), async () => {
    await delay(40);
    return HttpResponse.json({ markets: Array.from(STATE.markets.values()).map((m) => m.market) });
  }),

  http.get(u("/v1/orderbook/:marketId"), async ({ params }) => {
    const id = params.marketId as string;
    if (!known(id)) return HttpResponse.json({ error: { code: "MARKET_INACTIVE", message: id } }, { status: 404 });
    await delay(20);
    return HttpResponse.json(snapshotOf(id));
  }),

  http.get(u("/v1/trades/:marketId"), async ({ params, request }) => {
    const id = params.marketId as string;
    if (!known(id)) return HttpResponse.json({ error: { code: "MARKET_INACTIVE", message: id } }, { status: 404 });
    const url = new URL(request.url);
    const limit = Number(url.searchParams.get("limit") ?? 50);
    await delay(30);
    const trades = recentPublicTrades(id, limit);
    const nextBefore = trades.length ? trades[trades.length - 1].tsNs : null;
    return HttpResponse.json({ trades, nextBefore });
  }),

  http.get(u("/v1/funding/:marketId"), async ({ params }) => {
    const id = params.marketId as string;
    if (!known(id)) return HttpResponse.json({ error: { code: "MARKET_INACTIVE", message: id } }, { status: 404 });
    await delay(30);
    return HttpResponse.json(fundingFrame(id));
  }),

  http.post(u("/v1/orders"), async ({ request }) => {
    const body = (await request.json()) as OrderRequest;
    if (!known(body.marketId))
      return HttpResponse.json({ error: { code: "MARKET_INACTIVE", message: body.marketId } }, { status: 400 });
    const orderId = `ord_${Math.random().toString(36).slice(2, 12)}`;
    const ts = nsStr(nowNs());
    const remaining = body.type === "market" ? "0" : body.qty;
    const order: OpenOrder = {
      id: orderId,
      marketId: body.marketId,
      side: body.side,
      type: body.type,
      price: body.price,
      qty: body.qty,
      remaining,
      tsNs: ts,
      clientOrderId: body.clientOrderId,
    };
    if (body.type === "limit") {
      STATE.openOrders = [order, ...STATE.openOrders];
    } else {
      // Market — immediate fill
      const fill: Fill = {
        id: `fill_${Math.random().toString(36).slice(2, 10)}`,
        orderId,
        marketId: body.marketId,
        side: body.side,
        price: body.price,
        qty: body.qty,
        feeUsdc: (Number(body.qty) * Number(body.price) * 0.0005).toFixed(2),
        role: "taker",
        tsNs: ts,
        txHash: `0x${Math.random().toString(16).slice(2, 10).padEnd(64, "0")}`,
      };
      pushFill(fill);
      // Apply naive position update — long if buy
      const m = STATE.markets.get(body.marketId)!;
      const existing = STATE.positions.find((p) => p.marketId === body.marketId);
      const dir = body.side === "buy" ? 1 : -1;
      const qty = Number(body.qty) * dir;
      if (existing) {
        const oldSize = Number(existing.size) * (existing.side === "long" ? 1 : -1);
        const newSize = oldSize + qty;
        if (Math.abs(newSize) < Number(m.market.lotSize)) {
          STATE.positions = STATE.positions.filter((p) => p.marketId !== body.marketId);
        } else {
          existing.size = Math.abs(newSize).toFixed(4);
          existing.side = newSize > 0 ? "long" : "short";
          existing.entryPriceX18 = body.price;
        }
      } else {
        STATE.positions.push({
          marketId: body.marketId,
          size: body.qty,
          side: body.side === "buy" ? "long" : "short",
          entryPriceX18: body.price,
          markPriceX18: body.price,
          notionalUsdc: (Number(body.price) * Number(body.qty)).toFixed(2),
          unrealisedPnlUsdc: "0.00",
          realisedPnlUsdc: "0.00",
          leverage: "2.0",
          liquidationPriceX18: (Number(body.price) * (body.side === "buy" ? 0.92 : 1.08)).toFixed(2),
          fundingPaidUsdc: "0.00",
          lastUpdatedTsNs: ts,
        });
      }
      refreshMarks();
    }
    await delay(60);
    return HttpResponse.json({ orderId, status: "accepted", tsNs: ts });
  }),

  http.delete(u("/v1/orders/:orderId"), async ({ params }) => {
    const id = params.orderId as string;
    STATE.openOrders = STATE.openOrders.filter((o) => o.id !== id);
    await delay(40);
    return HttpResponse.json({ orderId: id, status: "cancelled" });
  }),

  http.get(u("/v1/orders/open"), async () => {
    await delay(30);
    return HttpResponse.json({ orders: STATE.openOrders });
  }),

  http.get(u("/v1/positions"), async () => {
    refreshMarks();
    await delay(30);
    return HttpResponse.json(positionsFrame());
  }),

  http.get(u("/v1/fills"), async ({ request }) => {
    const url = new URL(request.url);
    const marketId = url.searchParams.get("marketId");
    const limit = Number(url.searchParams.get("limit") ?? 100);
    let fills = STATE.fills;
    if (marketId) fills = fills.filter((f) => f.marketId === marketId);
    const slice = fills.slice(0, limit);
    const nextBefore = slice.length ? slice[slice.length - 1].tsNs : null;
    await delay(30);
    return HttpResponse.json({ fills: slice, nextBefore });
  }),

  http.get(u("/v1/account/balance"), async () => {
    await delay(30);
    return HttpResponse.json(balanceFrame());
  }),

  http.post(u("/v1/auth/siwe/nonce"), async () => {
    const nonce = Math.random().toString(36).slice(2, 18);
    return HttpResponse.json({ nonce });
  }),

  http.post(u("/v1/auth/siwe/verify"), async () => {
    const jwt = `jwt_${Math.random().toString(36).slice(2, 20)}`;
    STATE.jwt = jwt;
    return HttpResponse.json({ jwt, expiresAt: nsStr(nowNs() + 3_600_000_000_000n) });
  }),
];

File: crates/perplex-edge/src/handlers.rs
//! Endpoint implementations. Each function maps 1-1 to a section of api-contract.md.
//! Numeric values cross the wire as decimal strings (never floats) per section 0 of the contract.

use axum::extract::{Path, Query, State};
use axum::Json;
use rust_decimal::Decimal;
use serde::Deserialize;

use crate::auth::{issue_jwt, verify_siwe, AuthedUser};
use crate::error::ApiError;
use crate::state::{now_ns, AppState};
use crate::types::*;

const MAX_TRADES: usize = 1000;
const MAX_FILLS: usize = 1000;

// -- 1.1 GET /v1/markets ------------------------------------------------------

#[utoipa::path(get, path = "/v1/markets", responses((status = 200, body = MarketsResponse)))]
pub async fn list_markets(
    State(state): State<AppState>,
) -> Result<Json<MarketsResponse>, ApiError> {
    Ok(Json(MarketsResponse {
        markets: state.list_markets(),
    }))
}

// -- 1.2 GET /v1/orderbook/{marketId} -----------------------------------------

#[derive(Debug, Deserialize)]
pub struct OrderbookQuery {
    /// Optional level cap; ignored when "full".
    depth: Option<String>,
}

#[utoipa::path(
    get,
    path = "/v1/orderbook/{marketId}",
    params(("marketId" = String, Path, description = "Market id, e.g. btc-usd")),
    responses((status = 200, body = OrderbookSnapshot))
)]
pub async fn get_orderbook(
    State(state): State<AppState>,
    Path(market_id): Path<String>,
    Query(q): Query<OrderbookQuery>,
) -> Result<Json<OrderbookSnapshot>, ApiError> {
    if state.market(&market_id).is_none() {
        return Err(ApiError::NotFound);
    }
    let snap = state.orderbook(&market_id).unwrap_or_default();
    let depth = match q.depth.as_deref() {
        Some("full") | None => usize::MAX,
        Some("50") => 50,
        Some("200") => 200,
        Some(other) => other.parse::<usize>().unwrap_or(50),
    };
    let trim = |levels: Vec<[String; 2]>| levels.into_iter().take(depth).collect::<Vec<_>>();
    Ok(Json(OrderbookSnapshot {
        market_id,
        sequence: snap.sequence,
        bids: trim(snap.bids),
        asks: trim(snap.asks),
        ts_ns: now_ns().to_string(),
    }))
}

// -- 1.3 GET /v1/trades/{marketId} --------------------------------------------

#[derive(Debug, Deserialize)]
pub struct TradesQuery {
    limit: Option<usize>,
    before: Option<String>,
}

#[utoipa::path(
    get,
    path = "/v1/trades/{marketId}",
    params(("marketId" = String, Path, description = "Market id")),
    responses((status = 200, body = TradesResponse))
)]
pub async fn get_trades(
    State(state): State<AppState>,
    Path(market_id): Path<String>,
    Query(q): Query<TradesQuery>,
) -> Result<Json<TradesResponse>, ApiError> {
    if state.market(&market_id).is_none() {
        return Err(ApiError::NotFound);
    }
    let limit = q.limit.unwrap_or(100).min(MAX_TRADES);
    let mut trades = state.public_trades(&market_id, limit);
    if let Some(before) = q.before {
        trades.retain(|t| t.ts_ns < before);
    }
    let next_before = trades.last().map(|t| t.ts_ns.clone());
    Ok(Json(TradesResponse {
        trades,
        next_before,
    }))
}

// -- 1.4 GET /v1/funding/{marketId} -------------------------------------------

#[derive(Debug, Deserialize)]
pub struct FundingQuery {
    range: Option<String>,
}

#[utoipa::path(
    get,
    path = "/v1/funding/{marketId}",
    params(("marketId" = String, Path, description = "Market id")),
    responses((status = 200, body = FundingResponse))
)]
pub async fn get_funding(
    State(state): State<AppState>,
    Path(market_id): Path<String>,
    Query(q): Query<FundingQuery>,
) -> Result<Json<FundingResponse>, ApiError> {
    if state.market(&market_id).is_none() {
        return Err(ApiError::NotFound);
    }
    let window_ns: u64 = match q.range.as_deref() {
        Some("1h") => 3_600 * 1_000_000_000,
        Some("7d") => 7 * 86_400 * 1_000_000_000,
        _ => 24 * 3_600 * 1_000_000_000,
    };
    let cutoff = now_ns().saturating_sub(window_ns);
    let history: Vec<_> = state
        .funding_history(&market_id)
        .into_iter()
        .filter(|(t, _)| *t >= cutoff)
        .map(|(t, r)| FundingHistoryPoint {
            ts_ns: t.to_string(),
            rate_bps: r,
        })
        .collect();
    let current = history.last().map(|p| p.rate_bps).unwrap_or(0.0);
    Ok(Json(FundingResponse {
        market_id,
        current_rate_bps: current,
        next_settlement_ts_ns: (now_ns() + 8 * 3600 * 1_000_000_000).to_string(),
        history,
    }))
}

// -- 1.5 POST /v1/orders ------------------------------------------------------

#[utoipa::path(
    post,
    path = "/v1/orders",
    request_body = PlaceOrderRequest,
    responses(
        (status = 200, body = PlaceOrderResponse),
        (status = 400, description = "validation failure"),
        (status = 401, description = "auth failure")
    )
)]
pub async fn place_order(
    State(state): State<AppState>,
    user: AuthedUser,
    Json(req): Json<PlaceOrderRequest>,
) -> Result<Json<PlaceOrderResponse>, ApiError> {
    if state.market(&req.market_id).is_none() {
        return Err(ApiError::MarketInactive);
    }
    if req.side != "buy" && req.side != "sell" {
        return Err(ApiError::BadRequest("side must be buy or sell".into()));
    }
    if req.order_type != "limit" && req.order_type != "market" {
        return Err(ApiError::BadRequest("type must be limit or market".into()));
    }
    if !matches!(req.time_in_force.as_str(), "gtc" | "ioc" | "fok") {
        return Err(ApiError::BadRequest("invalid timeInForce".into()));
    }
    if !req.signature.starts_with("0x") || req.signature.len() < 132 {
        return Err(ApiError::InvalidSignature);
    }
    let limit_price = if req.order_type == "limit" {
        Some(
            req.price
                .clone()
                .ok_or_else(|| ApiError::BadRequest("limit requires price".into()))?,
        )
    } else {
        None
    };

    let taker_qty: Decimal = req
        .qty
        .parse()
        .map_err(|_| ApiError::BadRequest("invalid qty".into()))?;
    if taker_qty.is_zero() {
        return Err(ApiError::BadRequest("qty must be > 0".into()));
    }
    let taker_price: Option<Decimal> = match &limit_price {
        Some(p) => Some(
            p.parse()
                .map_err(|_| ApiError::BadRequest("invalid price".into()))?,
        ),
        None => None,
    };

    let order_id = format!("ord_{}", ulid::Ulid::new());
    let ts_ns = now_ns().to_string();

    // Cross against the resting book. Maker orders are decremented in place and
    // the public orderbook snapshot is rebuilt by `match_taker`.
    let matches = state.match_taker(&req.market_id, &req.side, taker_price, taker_qty);
    let filled_qty: Decimal = matches.iter().map(|m| m.qty).sum();
    let taker_remaining = taker_qty - filled_qty;
    let taker_side = req.side.clone();
    let maker_side = if taker_side == "buy" { "sell" } else { "buy" };

    for m in &matches {
        let price_str = m.price.normalize().to_string();
        let qty_str = m.qty.normalize().to_string();
        let ts_str = m.ts_ns.to_string();

        state.record_fill(
            &user.address,
            FillInfo {
                id: format!("fil_{}", ulid::Ulid::new()),
                order_id: order_id.clone(),
                market_id: req.market_id.clone(),
                side: taker_side.clone(),
                price: price_str.clone(),
                qty: qty_str.clone(),
                fee_usdc: "0".into(),
                role: "taker".into(),
                ts_ns: ts_str.clone(),
                tx_hash: String::new(),
            },
        );
        state.record_fill(
            &m.maker_address,
            FillInfo {
                id: format!("fil_{}", ulid::Ulid::new()),
                order_id: m.maker_order_id.clone(),
                market_id: req.market_id.clone(),
                side: maker_side.to_string(),
                price: price_str.clone(),
                qty: qty_str.clone(),
                fee_usdc: "0".into(),
                role: "maker".into(),
                ts_ns: ts_str.clone(),
                tx_hash: String::new(),
            },
        );
        state.record_trade(
            &req.market_id,
            PublicTrade {
                id: format!("trd_{}", ulid::Ulid::new()),
                market_id: req.market_id.clone(),
                price: price_str,
                qty: qty_str,
                side: taker_side.clone(),
                ts_ns: ts_str,
            },
        );

        state.upsert_position(&user.address, &req.market_id, &taker_side, m.qty, m.price);
        state.upsert_position(&m.maker_address, &req.market_id, maker_side, m.qty, m.price);
    }

    // Rest leftover liquidity for limit orders only. Market orders with a
    // remainder are dropped (no resting at price 0).
    if !taker_remaining.is_zero() && req.order_type == "limit" {
        let price = limit_price.clone().unwrap_or_else(|| "0".to_string());
        state.add_open_order(
            &user.address,
            OpenOrder {
                id: order_id.clone(),
                market_id: req.market_id.clone(),
                side: req.side.clone(),
                order_type: req.order_type.clone(),
                price,
                qty: req.qty.clone(),
                remaining: taker_remaining.normalize().to_string(),
                ts_ns: ts_ns.clone(),
                client_order_id: req.client_order_id.clone(),
            },
        );
    }

    let status = if !matches.is_empty() && taker_remaining.is_zero() {
        "filled"
    } else if !matches.is_empty() {
        "partial"
    } else {
        "accepted"
    };

    Ok(Json(PlaceOrderResponse {
        order_id,
        status: status.into(),
        ts_ns,
    }))
}

// -- 1.6 DELETE /v1/orders/{orderId} ------------------------------------------

#[utoipa::path(
    delete,
    path = "/v1/orders/{orderId}",
    params(("orderId" = String, Path, description = "Order id")),
    responses((status = 200, body = CancelOrderResponse))
)]
pub async fn cancel_order(
    State(state): State<AppState>,
    user: AuthedUser,
    Path(order_id): Path<String>,
) -> Result<Json<CancelOrderResponse>, ApiError> {
    // Idempotent: re-cancel returns 200 with status "cancelled" regardless of prior state.
    let _ = state.cancel_order(&user.address, &order_id);
    Ok(Json(CancelOrderResponse {
        order_id,
        status: "cancelled".into(),
    }))
}

// -- 1.7 GET /v1/orders/open --------------------------------------------------

#[utoipa::path(get, path = "/v1/orders/open", responses((status = 200, body = OpenOrdersResponse)))]
pub async fn list_open_orders(
    State(state): State<AppState>,
    user: AuthedUser,
) -> Result<Json<OpenOrdersResponse>, ApiError> {
    Ok(Json(OpenOrdersResponse {
        orders: state.open_orders_for(&user.address),
    }))
}

// -- 1.8 GET /v1/positions ----------------------------------------------------

#[utoipa::path(get, path = "/v1/positions", responses((status = 200, body = PositionsResponse)))]
pub async fn list_positions(
    State(state): State<AppState>,
    user: AuthedUser,
) -> Result<Json<PositionsResponse>, ApiError> {
    let positions = state.positions_for(&user.address);
    let collateral = state.vault_balance(&user.address);
    Ok(Json(PositionsResponse {
        collateral_usdc: collateral.clone(),
        free_collateral_usdc: collateral,
        total_unrealised_pnl_usdc: "0".into(),
        total_notional_usdc: "0".into(),
        positions,
    }))
}

// -- 1.9 GET /v1/fills --------------------------------------------------------

#[derive(Debug, Deserialize)]
pub struct FillsQuery {
    #[serde(rename = "marketId")]
    market_id: Option<String>,
    limit: Option<usize>,
    before: Option<String>,
}

#[utoipa::path(get, path = "/v1/fills", responses((status = 200, body = FillsResponse)))]
pub async fn list_fills(
    State(state): State<AppState>,
    user: AuthedUser,
    Query(q): Query<FillsQuery>,
) -> Result<Json<FillsResponse>, ApiError> {
    let limit = q.limit.unwrap_or(100).min(MAX_FILLS);
    let mut fills = state.fills_for(&user.address, limit);
    if let Some(mid) = q.market_id {
        fills.retain(|f| f.market_id == mid);
    }
    if let Some(before) = q.before {
        fills.retain(|f| f.ts_ns < before);
    }
    let next_before = fills.last().map(|f| f.ts_ns.clone());
    Ok(Json(FillsResponse { fills, next_before }))
}

// -- 1.10 SIWE flow -----------------------------------------------------------

#[utoipa::path(post, path = "/v1/auth/siwe/nonce", request_body = SiweNonceRequest, responses((status = 200, body = SiweNonceResponse)))]
pub async fn siwe_nonce(
    State(state): State<AppState>,
    Json(req): Json<SiweNonceRequest>,
) -> Result<Json<SiweNonceResponse>, ApiError> {
    if !req.address.starts_with("0x") || req.address.len() != 42 {
        return Err(ApiError::BadRequest("invalid address".into()));
    }
    let nonce = state.issue_siwe_nonce(&req.address);
    Ok(Json(SiweNonceResponse { nonce }))
}

#[utoipa::path(post, path = "/v1/auth/siwe/verify", request_body = SiweVerifyRequest, responses((status = 200, body = SiweVerifyResponse)))]
pub async fn siwe_verify(
    State(state): State<AppState>,
    Json(req): Json<SiweVerifyRequest>,
) -> Result<Json<SiweVerifyResponse>, ApiError> {
    let address = verify_siwe(&req.message, &req.signature, &state)?;
    let (jwt, exp_secs) = issue_jwt(state.jwt_secret(), &address)?;
    Ok(Json(SiweVerifyResponse {
        jwt,
        expires_at: (exp_secs as u128 * 1_000_000_000).to_string(),
    }))
}

// -- 1.11 GET /v1/account/balance ---------------------------------------------

#[utoipa::path(get, path = "/v1/account/balance", responses((status = 200, body = BalanceResponse)))]
pub async fn get_balance(
    State(state): State<AppState>,
    user: AuthedUser,
) -> Result<Json<BalanceResponse>, ApiError> {
    Ok(Json(BalanceResponse {
        vault_balance_usdc: state.vault_balance(&user.address),
        wallet_usdc_balance: "0".into(),
        pending_deposits: vec![],
        pending_withdrawals: vec![],
    }))
}

// -- dev helper: mint a JWT for `address` without going through SIWE. Returns the same JSON
//    shape as `/v1/auth/siwe/verify` so scripts can use a single parser for both. Exposed
//    only by `build_router_with_dev_token` (integration tests + local dev). Never wired into
//    the production router.
pub async fn dev_token(
    State(state): State<AppState>,
    Path(address): Path<String>,
) -> Result<Json<SiweVerifyResponse>, ApiError> {
    let (jwt, exp_secs) = issue_jwt(state.jwt_secret(), &address)?;
    Ok(Json(SiweVerifyResponse {
        jwt,
        expires_at: (exp_secs as u128 * 1_000_000_000).to_string(),
    }))
}
